pkg/virt-controller/watch/pool/pool.go

```diff
@@ -889,9 +917,55 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
 
 func (c *Controller) proactiveUpdate(pool *poolv1.VirtualMachinePool, vmUpdatedList []*virtv1.VirtualMachine) error {
 	var wg sync.WaitGroup
-	wg.Add(len(vmUpdatedList))
+
+	maxUnavailable := intstr.FromString("25%")
+	if pool.Spec.MaxUnavailable != nil {
+		maxUnavailable = *pool.Spec.MaxUnavailable
+	}
+
+	totalReplicas := int32(1)
+	if pool.Spec.Replicas != nil {
+		totalReplicas = *pool.Spec.Replicas
+	}
+
+	var maxUnavailableInt int
+	if maxUnavailable.Type == intstr.String {
+		percentage, err := strconv.ParseInt(strings.TrimSuffix(maxUnavailable.StrVal, "%"), 10, 32)
+		if err != nil {
+			return fmt.Errorf("invalid maxUnavailable percentage: %v", err)
+		}
+		maxUnavailableInt = int(float64(totalReplicas) * float64(percentage) / 100.0)
+	} else {
+		maxUnavailableInt = int(maxUnavailable.IntVal)
+	}
+
+	if maxUnavailableInt < 1 {
+		maxUnavailableInt = 1
+	}
+
 	errChan := make(chan error, len(vmUpdatedList))
+	unavailableCount, err := c.getUnavailableVMICount(pool, vmUpdatedList)
+	if err != nil {
+		errChan <- err
+		return err
+	}
+
+	if unavailableCount > 0 {
+		return fmt.Errorf("cannot proceed with proactive updates: %d VMIs are unavailable", unavailableCount)
+	}
+
 	for i := 0; i < len(vmUpdatedList); i++ {
+		unavailableCount, err := c.getUnavailableVMICount(pool, vmUpdatedList)
+		if err != nil {
+			errChan <- err
+			return err
```

Comment:
This for loop does not run in a separate routine. Why bother writing to the channel?